libsignal-service/src/sealed_session_cipher.rs

```diff
@@ -0,0 +1,1070 @@
+    ) -> Result<Vec<u8>, SealedSessionError> {
+        if ciphertext.len() < 10 {
+            return Err(SealedSessionError::InvalidMacError(
+                MacError::CiphertextNotLongEnough(ciphertext.len()),
+            ));
+        }
+
+        let (ciphertext_part1, their_mac) =
+            ciphertext.split_at(ciphertext.len() - 10);
+
+        let mut verifier = Hmac::<Sha256>::new_varkey(&mac_key)
+            .map_err(|_| MacError::InvalidKeyLength)?;
+        verifier.update(&ciphertext_part1);
+        let digest = verifier.finalize().into_bytes();
+        let our_mac = &digest[..10];
+
+        if our_mac != their_mac {
+            return Err(SealedSessionError::InvalidMacError(MacError::BadMac));
+        }
+
+        let mut decrypted = ciphertext_part1.to_vec();
+        let mut cipher = Aes256Ctr::new(cipher_key.into(), &[0u8; 16].into());
+        cipher.decrypt(&mut decrypted);
+
+        Ok(decrypted)
+    }
+
+    fn get_preferred_address(
+        &self,
+        certificate: &SenderCertificate,
+    ) -> Result<Address, SealedSessionError> {
+        if let Some(ref sender_uuid) = certificate.sender_uuid {
+            let address =
+                Address::new(sender_uuid, certificate.sender_device_id as i32);
+            if self.store_context.contains_session(&address)? {
+                return Ok(address);
+            }
+        } else if let Some(ref sender_e164) = certificate.sender_e164 {
+            let address =
+                Address::new(sender_e164, certificate.sender_device_id as i32);
+            if self.store_context.contains_session(&address)? {
+                return Ok(address);
+            }
+        }
+
+        Ok(Address::new(
+            certificate.sender(),
+            certificate.sender_device_id as i32,
+        ))
+    }
+
+    fn decrypt_message_content(
+        &self,
+        message: &UnidentifiedSenderMessageContent,
+    ) -> Result<(Vec<u8>, u32), SealedSessionError> {
+        let sender = self.get_preferred_address(&message.sender_certificate)?;
+        let session_cipher =
+            SessionCipher::new(&self.context, &self.store_context, &sender)?;
+        let msg = match message.r#type {
+            CiphertextType::Signal => {
+                let msg = session_cipher.decrypt_message(
+                    &SignalMessage::deserialize(
+                        &self.context,
+                        &message.content,
+                    )?,
+                )?;
+                msg.as_slice().to_vec()
+            }
+            CiphertextType::PreKey => {
+                let msg = session_cipher.decrypt_pre_key_message(
+                    &PreKeySignalMessage::deserialize(
+                        &self.context,
+                        &message.content,
+                    )?,
+                )?;
+                msg.as_slice().to_vec()
+            }
+            _ => unreachable!("unknown message from unidentified sender type"),
```

Comment:
Shouldn't this  be an error instead of a panic, to be forward-compatible?

Reply:
well, it has been tested in the caller, it is an assert - it means we really mis-used the API in this case, so I'd still panic.